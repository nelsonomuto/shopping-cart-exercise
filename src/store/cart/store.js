const LOCALSTORAGE_CART_KEY = "rtg-cart"

const setCartItemsLocalStorage = cartItems => {
  localStorage.setItem(LOCALSTORAGE_CART_KEY, JSON.stringify(cartItems))
}

const getCartItemsLocalStorage = () => {
  if (!localStorage.getItem(LOCALSTORAGE_CART_KEY)) {
    setCartItemsLocalStorage([])
  }

  return JSON.parse(localStorage.getItem(LOCALSTORAGE_CART_KEY))
}

const store = {
  get: getCartItemsLocalStorage,
  set: setCartItemsLocalStorage,
}

export default store
