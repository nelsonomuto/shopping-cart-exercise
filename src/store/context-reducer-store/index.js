import React from "react"
import createPersistentStore from "./persistentStore"

export default function createContextStore(
  name,
  reducer,
  defaultState,
  options = {}
) {
  // Create a "global" context to store our redux state in
  // this is private, we abstract this away from consumer
  const StoreContext = React.createContext()
  const _store = createPersistentStore(name, defaultState)

  // Set up serializers/deserializers for the store (if you only want to
  // stash part of it, or if you want to encrypt it)
  const self = content => content
  const serialize = options.serializeStore || self
  const deserialize = options.deserializeStore || self
  const store = { clear: _store.clear }
  store.set = content => _store.set(serialize(content))
  store.get = () => deserialize(_store.get())

  // Use this to wrap the whole portion of the app that needs access to the Store
  const StoreProvider = ({ children }) => {
    const initialState = store.get()
    const [state, dispatch] = React.useReducer(reducer, initialState)
    return (
      <StoreContext.Provider value={{ state, dispatch }}>
        {children}
      </StoreContext.Provider>
    )
  }

  // Use this to access state and dispatch action, like useReducer
  // except it is available locally anywhere under the StoreProvider wrapper
  const useStore = () => {
    const { state, dispatch } = React.useContext(StoreContext)

    React.useEffect(() => {
      store.set(state)
    }, [state])

    return [state, dispatch]
  }

  return {
    store,
    useStore,
    StoreProvider,
  }
}
