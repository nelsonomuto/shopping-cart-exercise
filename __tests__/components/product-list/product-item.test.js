import { mount, shallow } from "enzyme"
import React from "react"
import ProductItem from "../../../src/components/product/product-item"
import store from "../../../src/store/cart/store"

const mockCartItems = [
  {
    sku: "fake-sku",
    title: "fake-title",
    price: 999.99,
    image: "fake-image-url",
    qty: 1,
  },
]

const mockProduct = {
  sku: "fake-sku",
  title: "fake-title",
  price: 999.99,
  image: "fake-image-url",
}

describe("<ProductItem />", () => {
  afterEach(() => {
    store.clear()
  })

  it("renders product item from product", () => {
    const component = shallow(<ProductItem {...mockProduct} />)
    expect(component).toMatchSnapshot()
  })

  it("adds item to cart", () => {
    const component = mount(<ProductItem {...mockProduct} />)
    const addButton = component.find("#add-to-cart")
    addButton.simulate("click")
    expect(store.get()).toEqual(mockCartItems)
  })
})
