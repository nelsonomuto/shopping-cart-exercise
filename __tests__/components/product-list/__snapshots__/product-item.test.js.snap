// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductItem /> renders product item from product 1`] = `
<div
  className="product cell small-12 grid-x grid-margin-x"
  id="fake-sku"
>
  <div
    className="product-image cell small-2"
  >
    <img
      alt="fake-title"
      src="fake-image-url"
    />
  </div>
  <div
    className="product-title cell small-4"
  >
    fake-title
  </div>
  <div
    className="product-sku cell small-2"
  >
    fake-sku
  </div>
  <div
    className="product-price cell small-2"
  >
    $
    999.99
  </div>
  <div
    className="product-add-to-cart cell small-2"
  >
    <button
      id="add-to-cart"
      onClick={[Function]}
    >
       
      Add to Cart
    </button>
  </div>
</div>
`;
